Kashoo/synctos
# Issue 292: Do not depend on buggy Date object for date value comparisons

# Description

The Date object implementation provided by otto has become even less reliable in the version included in Sync Gateway 2.0.0 Beta 2. This change insulates a user of synctos from its buggy behaviour while continuing to provide the same features.

# Testing

`date` validation type:

1. Generate a sync function from the `date-doc-definitions.js` file: `./make-sync-function -j test/resources/date-doc-definitions.js build/output.txt`
2. Insert the generated sync function into a Sync Gateway configuration file
3. Create a series of documents to exercise the format validation, range and immutability constraints, following the examples from `test/date.spec.js`. For example, this should be created successfully:

```
PUT /test/dateDoc HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{
	"formatValidationProp": "2018-04-09"
}
```

`datetime` validation type:

1. Generate a sync function from the `datetime-doc-definitions.js` file: `./make-sync-function -j test/resources/datetime-doc-definitions.js build/output.txt`
2. Insert the generated sync function into a Sync Gateway configuration file
3. Create a series of documents to exercise the format validation, range and immutability constraints, following the examples from `test/datetime.spec.js`. For example, this should be created successfully:

```
PUT /test/datetimeDoc HTTP/1.1
Host: localhost:4985
Content-Type: application/json

{
	"formatValidationProp": "1999-12-31T24:00:00.000-08:00"
}
```

# Related Issue

* GitHub issue link: #292
